src/Razor/src/Microsoft.VisualStudio.LanguageServerClient.Razor/DefaultLSPDocumentManager.cs

```diff
@@ -0,0 +1,133 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Composition;
+using System.Diagnostics;
+using Microsoft.VisualStudio.Shell;
+using Microsoft.VisualStudio.Text;
+using Microsoft.VisualStudio.Text.Editor;
+using Microsoft.VisualStudio.Threading;
+
+namespace Microsoft.VisualStudio.LanguageServerClient.Razor
+{
+    [Shared]
+    [Export(typeof(LSPDocumentManager))]
+    internal class DefaultLSPDocumentManager : LSPDocumentManagerBase
+    {
+        private readonly JoinableTaskContext _joinableTaskContext;
+        private readonly FileUriProvider _fileUriProvider;
+        private readonly LSPDocumentFactory _documentFactory;
+        private readonly Dictionary<Uri, DocumentTracker> _documents;
+
+        [ImportingConstructor]
+        public DefaultLSPDocumentManager(
+            JoinableTaskContext joinableTaskContext,
+            FileUriProvider fileUriProvider,
+            LSPDocumentFactory documentFactory)
+        {
+            if (joinableTaskContext is null)
+            {
+                throw new ArgumentNullException(nameof(joinableTaskContext));
+            }
```

Comment:
Nit: have you guys looked at something like Contract.Requires<ArgumentNullException>(x != null, nameof(x));

Reply:
No not at all. Does that offer added benefits over manually checking null and throwing?

Reply:
Not really, just more compact syntax of writing the same thing. Not a big deal. 

---
In reply to: [381423703](https://github.com/dotnet/aspnetcore-tooling/pull/1602#discussion_r381423703) [](ancestors = 381423703)

Reply:
Alternatively, could just do something like:

_joinableTaskContext = joinableTaskContext ?? throw new ArgumentNullException(nameof(joinableTaskContext));

---
In reply to: [381429087](https://github.com/dotnet/aspnetcore-tooling/pull/1602#discussion_r381429087) [](ancestors = 381429087,381423703)

Reply:
Alternatively (without depending on another library for a null check), how do you feel about the syntax:
`joinableTaskContext = joinableTaskContext ?? throw new ArgumentNullException(nameof(joinableTaskContext))?`

Reply:
Ya we've considered doing the single line null-ref checks but in an effort to stay consistent with the repo/aspnet org have kept the multi-line versions.